From a64c0a969fdc25c312e06c6df9af01ac1229e1ef Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Tue, 29 Apr 2025 22:49:07 +0800
Subject: [PATCH 713/960] downstream: arm64: mpam: Restore the expected MPAM
 sysregs on cpuhp

When a CPU is brought online, only the MPAM1_EL1 sysreg is set. For
completeness restore the MPAM0_EL1 register too.

This isn't expected to cause a problem as cpuhp happens from the idle
thread, which can't return to user-space.

Signed-off-by: James Morse <james.morse@arm.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kernel/cpufeature.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/arch/arm64/kernel/cpufeature.c b/arch/arm64/kernel/cpufeature.c
index 5b574599ecde..ec96decc3e60 100644
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -2456,6 +2456,11 @@ static void __maybe_unused
 cpu_enable_mpam(const struct arm64_cpu_capabilities *entry)
 {
 	u64 idr = read_sanitised_ftr_reg(SYS_MPAMIDR_EL1);
+	int cpu = smp_processor_id();
+	u64 regval = 0;
+
+	if (IS_ENABLED(CONFIG_ARM64_MPAM))
+		regval = READ_ONCE(per_cpu(arm64_mpam_current, cpu));
 
 	/*
 	 * Initialise MPAM EL2 registers and disable EL2 traps.
@@ -2472,6 +2477,7 @@ cpu_enable_mpam(const struct arm64_cpu_capabilities *entry)
 	 * been throttled to release the lock.
 	 */
 	write_sysreg_s(0, SYS_MPAM1_EL1);
+	write_sysreg_s(regval, SYS_MPAM0_EL1);
 }
 
 static void mpam_extra_caps(void)
-- 
2.43.0

